keccak-hasher/src/lib.rs

```diff
@@ -2,50 +2,19 @@
 //!
 //! [keccak]: https://keccak.team/keccak.html
 #![cfg_attr(docsrs, feature(doc_auto_cfg))]
-#![cfg_attr(feature = "frozen-abi", feature(min_specialization))]
 #![no_std]
-#[cfg(feature = "std")]
-extern crate std;
 
-#[cfg(any(feature = "sha3", not(target_os = "solana")))]
+#[cfg(all(feature = "offchain", not(target_os = "solana")))]
 use sha3::{Digest, Keccak256};
-pub use solana_hash::{ParseHashError, HASH_BYTES, MAX_BASE58_LEN};
-#[cfg(feature = "borsh")]
-use {
-    borsh::{BorshDeserialize, BorshSchema, BorshSerialize},
-    std::string::ToString,
-};
-use {
-    core::{fmt, str::FromStr},
-    solana_sanitize::Sanitize,
-};
+pub use solana_hash::{Hash, ParseHashError, HASH_BYTES, MAX_BASE58_LEN};
 
-// TODO: replace this with `solana_hash::Hash` in the
-// next breaking change.
-// It's a breaking change because the field is public
-// here and private in `solana_hash`, and making
-// it public in `solana_hash` would break wasm-bindgen
-#[cfg_attr(feature = "frozen-abi", derive(solana_frozen_abi_macro::AbiExample))]
-#[cfg_attr(
-    feature = "borsh",
-    derive(BorshSerialize, BorshDeserialize, BorshSchema),
-    borsh(crate = "borsh")
-)]
-#[cfg_attr(
-    feature = "serde",
-    derive(serde_derive::Deserialize, serde_derive::Serialize)
-)]
-#[derive(Clone, Copy, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
-#[repr(transparent)]
-pub struct Hash(pub [u8; HASH_BYTES]);
-
-#[cfg(any(feature = "sha3", not(target_os = "solana")))]
 #[derive(Clone, Default)]
+#[cfg(all(feature = "offchain", not(target_os = "solana")))]
 pub struct Hasher {
     hasher: Keccak256,
 }
 
-#[cfg(any(feature = "sha3", not(target_os = "solana")))]
+#[cfg(all(feature = "offchain", not(target_os = "solana")))]
```

Comment:
I am bit confused about the "offchain" feature. It seems highly correlated with `not(target_os = "solana")`. For example, when `not(target_os = "solana")`, you need to have the "offchain" flag enabled, otherwise you get "nothing". So it seems that having  `not(target_os = "solana")` should be enough? I get the feeling that I am missing something.

Reply:
The concept is to make these kinds of libraries usable in pinocchio environments, where people want to minimize dependencies in their build.

Currently, things like `create_program_address` just fail in non-solana builds: https://github.com/anza-xyz/pinocchio/blob/3fecb3e569442e219185e5fe35854bfbf971e23a/sdk/pinocchio/src/pubkey.rs#L203

If we want pinocchio users to use these crates too, we should offer the opportunity to opt in of having these functions work if `target_os != solana`. This way, pinocchio can just not enable the `offchain` feature and never pull in the crates that implement the offchain behavior. Does that make sense?

Reply:
Yep, totally. Took me a second look to figure out the logic. 😅